Prevent negative quantity

lomamech opened the pull request:
Prevent negative quantity in client side in
- All stock exit
- Inventory adjustement
- Aggregate stock consumption

test(Prevent Negative quantity)
- Add integration test for prevent negative quantity during distribution
  lots to patient, aggregate consumption and adjustement inventories
closes #5205

lomamech committed:
Improvement: get lot in stock and by depot

lomamech committed:
Check if stock movements, which may overconsume the quantity in stock and generate negative quantity in stock

lomamech committed:
Use error badRequest

lomamech committed:
feature(Prevent negative quantity)
Prevent negative quantity in client side in
- All stock exit
- Inventory adjustement
- Aggregate stock consumption

lomamech committed:
test(Prevent negative quantity)
- Implement E2E test for prevent negative quantity

lomamech committed:
Test: restaure value for stock tests E2E

lomamech committed:
Fix and save conflict with branch master

jmcameron commented on client/src/i18n/en/errors.json:
```diff
@@ -25,6 +25,7 @@
     "ER_BAD_FIELD_ERROR" : "Column does not exist in database.",
     "ER_ROW_IS_REFERENCED" : "Cannot delete entity because entity is used in another table.",
     "ER_ROW_IS_REFERENCED_2" : "Cannot delete entity because entity is used in another table.",
+    "ER_PREVENT_NEGATIVE_QUANTITY_IN_STOCK" : "There are stock movements, which may overconsume the quantity in stock and generate negative quantity in stock",
```
I see that this error message will be used in multiple contexts.  Ideally there would be a separate message for each place this issue can come up.  Eg, one for stock entry that starts with:   "The proposed stock entry may overconsume..."
and for stock adjustment:  "The proposed stock adjustment may overconsume..."
Also, if you feel it is appropriate, I suggest replacing "may" with "will".

Whether you decided to add more error messages or not, please remove the comma.
"There are stock movements which may..."

lomamech replied:
```` Also, what is the strategy for dealing with sites that already have negative quantities? ````
In this case alone the positive adjustment is required, I just tested it with the database of IMCK and it works well

lomamech replied:
```` Also, what is the strategy for dealing with sites that already have negative quantities? ````
In this case alone the positive adjustment is required, I just tested it with the database of IMCK and it works well

jmcameron replied:
Once the more detailed error message is display, I agree that we don't need more detail in the temporary warning at the top of the page.

jmcameron commented on client/src/modules/stock/aggregated_consumption/aggregated_consumption.js:
```diff
@@ -228,6 +233,21 @@ function StockAggregatedConsumptionController(
+        vm.currentInventories = lots.filter(item => item.quantity > 0);
+
+        // Here we check directly if a Depot has inventories in stock available in current date
+        vm.emptyCurrentStock = !vm.currentInventories.length;
```
I think something like this would more readable:

vm.emptyCurrentStock = vm.currentInventories.length <= 0;

lomamech replied:
I remove **vm.emptyCurrentStock** this element, I did not use it anywhere

jmcameron replied:
Nice!

jmcameron commented on client/src/modules/stock/aggregated_consumption/aggregated_consumption.js:
```diff
@@ -249,6 +269,37 @@ function StockAggregatedConsumptionController(
+        };
+      });
+
+      Notify.danger('ERRORS.ER_PREVENT_NEGATIVE_QUANTITY_IN_STOCK');
```
This error message gives no clue regarding which stock item is causing the problem.  I think it would be helpful to give the stock item name/description in the error message.   Unless the next error message has those details?

lomamech replied:
Indeed there is a message that describes the products that are problematic
![image](https://user-images.githubusercontent.com/8035588/112827194-ead7e680-9085-11eb-9303-6d5171fbf0f7.png)

jmcameron replied:
Good to see this message with the details.

jmcameron commented on client/src/modules/stock/exit/exit.js:
```diff
@@ -29,21 +29,29 @@ function StockExitController(
     vm.depot = depot;
     loadInventories(vm.depot);
+    loadCurrentInventories(vm.depot);
+
```
How does this handle the case where the current inventories may already have negative values?

lomamech replied:
For products with negative quantities, the system will assimilate to products out of stock, which is why only those with positive quantities are filtered.
````
 function loadCurrentInventories(depot, dateTo = new Date()) {
    vm.loading = true;
    Stock.lots.read(null, { depot_uuid : depot.uuid, dateTo })
      .then(lots => {
        vm.currentInventories = lots.filter(item => item.quantity > 0);
      })
      .catch(Notify.handleError)
      .finally(() => {
        vm.loading = false;
      });
  }
````

jmcameron replied:
ok!

jmcameron replied:
ok. I understand!

jmcameron commented on client/src/modules/stock/inventory-adjustment/inventory-adjustment.js:
```diff
@@ -27,14 +27,23 @@ function StockInventoryAdjustmentController(
     vm.depot = depot;
     loadInventories(vm.depot);
+    loadCurrentInventories(vm.depot);
+
```
Same question as before: How does this handle cases where the already existing data has negatives?

lomamech replied:
Products with negative quantities will be considered as out of stock products.
![image](https://user-images.githubusercontent.com/8035588/112832518-570a1880-908d-11eb-858e-617d68c1cae9.png)

jmcameron replied:
Do we have a way to fix the lots with negative quantities?
Since lots with negative quantities will not be displayed, how do we find them and fix them?

lomamech replied:
![image](https://user-images.githubusercontent.com/8035588/112985356-fa276480-9157-11eb-8aee-a3b9baa8b1a0.png)
The stock adjustment module allows you to do this, you just have to check the option **Include exhausted or negative lots**

jmcameron replied:
Thanks for pointing this out.  I had forgotten about it.

jmcameron commented on server/controllers/stock/index.js:
```diff
@@ -284,14 +297,39 @@ async function createInventoryAdjustment(req, res, next) {
+
+    if (filteredInvalidData.length) {
+      throw new BadRequest(
+        `There are stock movements,
```
Are these duplicate error messages?

lomamech replied:
Yes, and it is so in several uses of  ````throw new BadRequest````
![image](https://user-images.githubusercontent.com/8035588/112832911-e1527c80-908d-11eb-9df9-0596aaac4b83.png)

jmcameron reviewed (commented):
It mostly looks good.  I was able to get the error message by doing a bogus stock inventory adjustment and I see the error message:
![image](https://user-images.githubusercontent.com/62145/112500381-8f37f100-8d45-11eb-84cc-a7dd4f0a9a71.png)

But when I scroll up, I see an unstranslated error message:
![image](https://user-images.githubusercontent.com/62145/112500498-abd42900-8d45-11eb-93b6-dc2c78bdbc10.png)

Also, please note that the Stock Inventory Adjustment page is too long.  I suggest making it shorter by default and require the user to scroll down only if many items need adjustment.  In production sites, you'll have to scroll anyway, so why make it long?

Also, what is the strategy for dealing with sites that already have negative quantities?

I was able to run the tests and they all ran fine except for the End-to-End tests (which is also failing for me on master).

lomamech committed:
improvement(Adjustement stock)
- Complete translation key for Adjustement of stock
- Reduice Inventory Adjustment page HEIGHT

mbayopanda commented on server/controllers/stock/index.js:
```diff
@@ -382,7 +422,43 @@ async function createMovement(req, res, next) {
+      }
+    });
+
+    filteredInvalidData = await params.lots.filter(l => l.quantity > l.quantityAvailable);
```
These checks in the server side is excellent 👍🏽

mbayopanda commented on server/controllers/stock/index.js:
```diff
@@ -382,7 +422,43 @@ async function createMovement(req, res, next) {
+        which may overconsume the quantity in stock and generate negative quantity in stock`,
+        `ERRORS.ER_PREVENT_NEGATIVE_QUANTITY_IN_STOCK`,
+      );
+    }
```
👍🏽

mbayopanda reviewed (approved):
Handling over consumption at the server-side is a good idea 👍🏽 , so I am waiting for all @jmcameron questions to be addressed to merge this.

lomamech committed:
Improvement: Clean and sanitaze unused element

lomamech committed:
improvement: Use specific error message for each cas of overconsumption of stocks

jmcameron reviewed (commented):
I tested it (based on your instructions) and it works fine.  Thanks for answering and addressing my previous comments.

I have two concerns:

1.  The warning message about problematic adjustment is hard to understand
     
     ![image](https://user-images.githubusercontent.com/62145/112891582-2924d400-908d-11eb-9599-ce3dccec077b.png)

     I would suggestion wording something like this (in English):

     Changing the quantity of [stock name] in lot label PL1 from 1000 to 0 will result in a negative inventory quantity.

     My guess is that the French text could also use improvement.

     I was not completely sure I understand what you are trying to say here, so take my advice with grain of salt (

2. When I tried doing a backdated Stock Exit, I ran into a problem when there was no (or negative) stock available on the specified date:
       
    ![image](https://user-images.githubusercontent.com/62145/112892592-5c1b9780-908e-11eb-9032-47f9bf329485.png)

    When I clicked on the [Add] button, nothing happened.  However when I look at the console, I see this error message:
    
    ![image](https://user-images.githubusercontent.com/62145/112892434-28407200-908e-11eb-906b-155bc7227348.png)

    We should handle this more gracefully.  I'm not sure whether this is related to the PR or not.  So if it makes sense to you we could make a new issue for it and handle it later.  Let me know what you want to do about this.

Also, I think that the length of time that the temporary warning message at the top of the screen says up is too short.   It is okay for success messages to stay up a short time, but warnings should stay up longer so people have time to read them!   I suggest something more than 5 seconds.

Finally, I think you are very close here, so I suggest re-basing after you do your final adjustments.

lomamech committed:
Improve warning message

lomamech committed:
Improve warning message

jmcameron reviewed (approved):
I think you have dealt with all the issues I raised.  I suggest re-basing and then releasing it.

jniles commented on client/src/modules/stock/exit/exit.js:
```diff
@@ -676,8 +734,11 @@ function StockExitController(
-        ReceiptModal.stockExitLossReceipt(document.uuid, bhConstants.flux.TO_LOSS);
-        reinit(form);
+
+        if (document.uuid) {
```
Why wouldn't there be a document uuid?

lomamech replied:
This could happen if two users connect to the Stock Exit interface and the first user consumes the entire stock of a product before the second validates its consumption, and as the control is also carried out at server level, an error message will be returned to the client without there being a document.uuid

jniles commented on client/src/modules/stock/inventory-adjustment/inventory-adjustment.js:
```diff
@@ -218,6 +239,36 @@ function StockInventoryAdjustmentController(
+      });
+
+      Notify.danger('ERRORS.ER_PREVENT_NEGATIVE_QUANTITY_IN_ADJUSTMENT_STOCK');
+      vm.$loading = false;
```
I think you mean `vm.loading` here.  I can't find another reference to `$loading` in this file.

jniles commented on server/controllers/stock/index.js:
```diff
@@ -379,7 +418,42 @@ async function createMovement(req, res, next) {
+    average_consumption_algo : params.average_consumption_algo,
+  };
+
+  stockAvailable = await core.getLotsDepot(null, paramsStock);
```
If you only use this check on exit, let's move this call into the `if (params.is_exit) {` black.  That way, we won't wait for this db request on stock entries.

jniles reviewed (changes requested):
I can't get this to work.  See below:

![meii5DGiss](https://user-images.githubusercontent.com/896472/113307364-f5e87c00-92fc-11eb-9fff-f4e6249c1423.gif)

Isn't this supposed to block me?

lomamech committed:
improvement(Negative quantity)
- Improvement of warning messages
- Improvement of the code by conditioning the download of available stock only in the event of Exit Stock

jniles reviewed (approved):
LGTM.  Great work!

bors r+
